src/util.js

```diff
@@ -45,16 +67,24 @@ export function messageHandlerFunc(context, sqsQueue, handler) {
     const errorWrap = context.service?.wrapError || context.gb?.wrapError || (e => e);
 
     let parsedMessage;
+    let decodedMessage;
+    let decodedRest;
     try {
+      const contentEncoding = rest.MessageAttributes?.['Content-Encoding']?.StringValue;
       parsedMessage = JSON.parse(Body);
+      if (contentEncoding === supportedCompression) {
+        decodedMessage = JSON.parse(inflateMessage(parsedMessage));
+        decodedRest = JSON.parse(JSON.stringify(rest)); // deep copy
+        delete decodedRest.MessageAttributes?.['Content-Encoding'];
+      }
     } catch (error) {
       logger.error('Failed to parse SQS Body as JSON', errorWrap(error));
       Object.defineProperty(error, ALREADY_LOGGED, { value: true, enumerable: false });
       safeEmit(sqsQueue.queueClient, 'error', callInfo);
       throw error;
     }
     try {
-      await handler(messageContext, parsedMessage, rest);
+      await handler(messageContext, decodedMessage || parsedMessage, decodedRest || rest);
```

Comment:
parsedMessage and rest should be the only variable passed in here.  The || makes it confusing.

Reply:
We need to retain both copies so we can publish it to the `*_unprocessable` queue if rejected by a service. However, we can definitely clean it up a bit